src/app/Events/EventTypes.tsx

```diff
@@ -0,0 +1,144 @@
+  const [perPage, setPerPage] = React.useState(10);
+  const [openRow, setOpenRow] = React.useState(-1);
+  const [filterText, setFilterText] = React.useState('');
+
+  const tableColumns = [
+    {
+      title: 'Name',
+      cellFormatters: [expandable]
+    },
+    'Type ID',
+    'Description',
+    'Categories'
+  ];
+
+  React.useEffect(() => {
+    const sub = context.commandChannel.onResponse('list-event-types')
+      .pipe(
+        filter(m => m.status === 0),
+        map(m => m.payload),
+      )
+      .subscribe(types => setTypes(types));
+    return () => sub.unsubscribe();
+  }, []);
+
+  React.useEffect(() => {
+    context.commandChannel.sendMessage('list-event-types');
+  }, []);
+
+  React.useEffect(() => {
+    const offset = (currentPage - 1) * perPage;
+    const page = filterTypesByText(types, filterText).slice(offset, offset + perPage);
+
+    const rows: any[] = [];
+    page.forEach((t: EventType, idx: number) => {
+      rows.push({ cells: [ t.name, t.typeId, t.description, getCategoryString(t) ], isOpen: (idx === openRow) });
+      if (idx === openRow) {
+        let child = '';
+        for (const opt in t.options) {
+          child += `${opt}=[${t.options[opt].defaultValue}]\t`;
+        }
+        rows.push({ parent: idx, fullWidth: true, cells: [ child ] });
+      }
+    });
+
+    setDisplayedTypes(rows);
+  }, [currentPage, perPage, types, openRow, filterText]);
+
+  const getCategoryString = (eventType: EventType): string => {
+    return eventType.category.join(', ').trim();
+  };
+
+  const filterTypesByText = (types, filter) => {
+    if (!filter) {
+      return types;
+    }
+    const includesSubstr = (a, b) => !!a && !!b && a.toLowerCase().includes(b.trim().toLowerCase());
+    return types.filter(t => {
+      if (includesSubstr(t.name, filter)) {
+        return true;
+      }
+      if (includesSubstr(t.typeId, filter)) {
+        return true;
+      }
+      if (includesSubstr(t.description, filter)) {
+        return true;
+      }
+      if (includesSubstr(getCategoryString(t), filter)) {
+        return true;
+      }
+      return false
+    });
+  };
+
+  const onCurrentPage = (evt, currentPage) => {
+    setOpenRow(-1);
+    setCurrentPage(currentPage);
+  };
+
+  const onPerPage = (evt, perPage) => setPerPage(perPage);
+
+  const onCollapse = (event, rowKey, isOpen) => {
+    if (isOpen) {
+      if (openRow === -1) {
+        setOpenRow(rowKey);
+      } else {
+        setOpenRow(rowKey > openRow ? rowKey - 1 : rowKey);
+      }
+    } else {
+      setOpenRow(-1);
+    }
+  };
+
+  // TODO replace table with data list so collapsed event options can be custom formatted
+  return(<>
+    <Split>
+      <SplitItem>
+        <TextInput name="eventFilter" id="eventFilter" type="search" placeholder="Filter..." aria-label="Event filter" onChange={setFilterText}/>
+      </SplitItem>
+      <SplitItem isFilled />
+      <SplitItem>
+        <Pagination
```

Comment:
The pagination code seems to mess up on specific switches between number of items to display.

E.g.

1. Default is 10. Page twice to items 21-30 out of 138
2. Switch to show 100. It states showing items 101-138 but no items are visible.

Reply:
Good catch. I'll work on this.

Reply:
Hmm. The issue seems to be that I'm trying to calculate the offset for which events to display on the single rendered page, and doing this reactively when any of the parameters (current page number, count per page, etc.) change. The problem is that when the changed parameter is the count per page we only know what the new value is, so the calculation of the current offset overshoots because it's already using the new count per page.

A simple solution is to simply reset the view to no offset (ie items 1-n) whenever the count per page is changed. The user will lose their place if they change the count per page after paging over. A more complex solution means adding some more state and calculation to figure out what changed, map the index of the previous first recording in view to its new index in the resized page, and shift to that offset. Do you think it's worthwhile implementing all that?

Reply:
To clarify, in the case of:

CPP = 10 : Page 2

When a user switches to CPP = 100, we don't know that the previous CPP was 10, just that we're on page 2?

Reply:
Pretty much. The rendered page is calculated in a React hook, and that hook is called whenever any of those parameters is updated. The hook also only knows the current values of each parameter. So if we're on Page=2 and CPP=10 and the user goes to the next page, we calculate an offset based off of Page=3 and CPP=10 and it works as expected. But if they change the CPP then we try to give them a view based on that CPP and the previous page number - if we have 138 items then Page=3 and CPP=100 doesn't make sense, it's beyond the range of the actual backing array, so we end up rendering nothing.

Reply:
Maybe we can hack it by checking if the offset exceeds the length of the backing array, and if so, set the current page number to the last one that would be in range?

Reply:
Is it possible to add state for a single int to store CPP as updates occur? Then if it exists, we can use it as the previous CPP to calculate the correct page.

Reply:
That would basically just be another `useState` hook like the existing one for CPP, except we would have to do something hackish like only set its value in the update hook, and make the update hook not depend on that parameter's value, even though it does actually use it. Not great IMO, and goes against the best practices recommended by React since we would be intentionally using a piece of state within the update hook and not updating when that piece of state changes.

Reply:
This sounds weird. It's natural for state change results depend on what the previous state was. Usually state change actions includes (from, to)...

Reply:
Not sure I followed the last line "not updating when that piece of state changes.". We'd be running the same update hook, so only acting when a piece of state changes, and in this hook we'd check the previous state stored, run our calculation, store the new state and continue as before.

Reply:
> This sounds weird. It's natural for state change results to depend on what the previous state was. Usually state change actions includes (from, to)...

Yea, you *can* get that with React state, but I would need to rewrite this entire component as a class component instead of a functional component.

> Not sure I followed the last line "not updating when that piece of state changes.". We'd be running the same update hook, so only acting when a piece of state changes, and in this hook we'd check the previous state stored, run our calculation, store the new state and continue as before.

Right, but then if the update hook depends on a piece of state and also updates that state, then you end up updating every time you update.